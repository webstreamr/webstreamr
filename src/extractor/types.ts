import { Context, UrlResult } from '../types';

export interface Extractor {
  readonly id: string;

  readonly label: string;

  readonly ttl: number;

  readonly supports: (url: URL) => boolean;

  readonly extract: (ctx: Context, url: URL, countryCode: string) => Promise<UrlResult>;
}
