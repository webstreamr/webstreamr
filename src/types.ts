import { Manifest, ManifestConfig } from 'stremio-addon-sdk';

export const TIMEOUT = 'timeout';

export interface Context {
  id: string;
  ip: string;
  config: Config;
  referer?: URL;
}

export type ManifestWithConfig = Manifest & { config: ManifestConfig[] };

export type Config = Partial<Record<CountryCode | 'excludeExternalUrls', string>>;

export type CountryCode = 'de' | 'en' | 'es' | 'fr' | 'it' | 'mx';

export type BlockedReason = 'cloudflare_challenge' | 'unknown';

export interface Meta {
  bytes?: number;
  countryCode: CountryCode;
  height?: number;
  title?: string;
}

export interface UrlResult {
  url: URL;
  isExternal?: boolean;
  error?: unknown;
  label: string;
  sourceId: string;
  ttl?: number;
  meta: Meta;
  requestHeaders?: Record<string, string>;
}
