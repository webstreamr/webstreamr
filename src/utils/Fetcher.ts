import { gunzipSync, gzipSync } from 'zlib';
import TTLCache from '@isaacs/ttlcache';
import { Mutex, Semaphore, SemaphoreInterface, withTimeout } from 'async-mutex';
import CachePolicy from 'http-cache-semantics';
import { LRUCache } from 'lru-cache';
import { Cookie, CookieJar } from 'tough-cookie';
import { fetch, Headers, RequestInit, Response } from 'undici';
import winston from 'winston';
import { BlockedError, HttpError, NotFoundError, QueueIsFullError, TimeoutError, TooManyRequestsError, TooManyTimeoutsError } from '../error';
import { BlockedReason, Context } from '../types';
import { envGet } from './env';

interface HttpCacheItem {
  policy: CachePolicy;
  status: number;
  statusText: string;
  body: string;
}

interface FlareSolverrResult {
  status: string;
  message: string;
  solution: {
    url: string;
    status: number;
    cookies: {
      domain: string;
      expires: number;
      httpOnly: boolean;
      name: string;
      path: string;
      sameSite: string;
      secure: boolean;
      value: string;
    }[];
    userAgent: string;
    headers: Record<string, string>;
    response: string;
  };
  startTimeStamp: number;
  endTimeStamp: number;
  version: string;
}

export type CustomRequestInit = RequestInit & {
  timeoutsCountThrow?: number;
  noFlareSolverr?: boolean;
  queueLimit?: number;
  queueTimeout?: number;
  timeout?: number;
};

export class Fetcher {
  private readonly MIN_CACHE_TTL = 900000; // 15m
  private readonly DEFAULT_TIMEOUT = 10000;
  private readonly DEFAULT_QUEUE_LIMIT = 5;
  private readonly DEFAULT_QUEUE_TIMEOUT = 5000;
  private readonly DEFAULT_FAILED_REQUEST_COUNT_THROW = 10;
  private readonly TIMEOUT_CACHE_TTL = 3600000; // 1h

  private readonly logger: winston.Logger;

  private readonly httpCache = new LRUCache<string, Buffer>({ max: 1024 });
  private readonly rateLimitedCache = new TTLCache<string, undefined>({ max: 1024 });
  private readonly semaphores = new Map<string, SemaphoreInterface>();
  private readonly hostUserAgentMap = new Map<string, string>();
  private readonly cookieJar = new CookieJar();

  private readonly timeoutsCountCache = new TTLCache<string, number>({ max: 1024 });
  private readonly timeoutsCountMutex = new Mutex();

  public constructor(logger: winston.Logger) {
    this.logger = logger;
  }

  public async text(ctx: Context, url: URL, init?: CustomRequestInit): Promise<string> {
    return (await this.cachedFetch(ctx, url, init)).body;
  };

  public async textPost(ctx: Context, url: URL, body: string, init?: CustomRequestInit): Promise<string> {
    return (await this.cachedFetch(ctx, url, { ...init, method: 'POST', body })).body;
  };

  public async head(ctx: Context, url: URL, init?: CustomRequestInit): Promise<CachePolicy.Headers> {
    return (await this.cachedFetch(ctx, url, { ...init, method: 'HEAD' })).policy.responseHeaders();
  };

  private getInit(ctx: Context, url: URL, init?: CustomRequestInit): CustomRequestInit {
    const cookieString = this.cookieJar.getCookieStringSync(url.href);

    return {
      ...init,
      headers: {
        'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
        'Accept-Language': 'en',
        'Priority': 'u=0',
        'User-Agent': this.hostUserAgentMap.get(url.host) ?? 'node',
        ...(cookieString && { Cookie: cookieString }),
        ...(ctx.ip && {
          'Forwarded': `for=${ctx.ip}`,
          'X-Forwarded-For': ctx.ip,
          'X-Forwarded-Proto': url.protocol.slice(0, -1),
          'X-Real-IP': ctx.ip,
        }),
        ...init?.headers,
      },
    };
  };

  private async handleHttpCacheItem(ctx: Context, httpCacheItem: HttpCacheItem, url: URL, init?: CustomRequestInit): Promise<HttpCacheItem> {
    if (httpCacheItem.status && httpCacheItem.status >= 200 && httpCacheItem.status <= 399) {
      return httpCacheItem;
    }

    if (httpCacheItem.status === 404) {
      throw new NotFoundError();
    }

    const responseHeaders = httpCacheItem.policy.responseHeaders();

    if (httpCacheItem.policy.responseHeaders()['cf-mitigated'] === 'challenge') {
      const noFlareSolverr = init?.noFlareSolverr ?? false;
      const flareSolverrEndpoint = envGet('FLARESOLVERR_ENDPOINT');
      if (noFlareSolverr || !flareSolverrEndpoint) {
        throw new BlockedError(BlockedReason.cloudflare_challenge, httpCacheItem.policy.responseHeaders());
      }

      this.logger.info(`Query FlareSolverr for ${url.href}`, ctx);

      const body = { cmd: 'request.get', url: url.href, session: 'default' };
      const challengeResult = await (await this.queuedFetch(ctx, new URL(flareSolverrEndpoint), { method: 'POST', body: JSON.stringify(body), headers: { 'Content-Type': 'application/json' }, queueLimit: 1, timeout: 15000 })).json() as FlareSolverrResult;
      if (challengeResult.status !== 'ok') {
        this.logger.warn(`FlareSolverr issue: ${JSON.stringify(challengeResult)}`, ctx);
        throw new BlockedError(BlockedReason.flaresolverr_failed, {});
      }

      challengeResult.solution.cookies.forEach((cookie) => {
        if (!['cf_clearance'].includes(cookie.name)) {
          return;
        }

        this.cookieJar.setCookie(
          new Cookie({
            key: cookie.name,
            value: cookie.value,
            expires: new Date(cookie.expires * 1000),
            domain: cookie.domain.replace(/^.+/, ''),
          }),
          url.href,
        );
      });

      this.hostUserAgentMap.set(url.host, challengeResult.solution.userAgent);

      httpCacheItem.body = challengeResult.solution.response;

      this.cacheSet(this.determineCacheKey(url, init), httpCacheItem);

      return httpCacheItem;
    }

    if (httpCacheItem.status === 403) {
      if (ctx.config.mediaFlowProxyUrl && url.href.startsWith(ctx.config.mediaFlowProxyUrl)) {
        throw new BlockedError(BlockedReason.media_flow_proxy_auth, httpCacheItem.policy.responseHeaders());
      }

      throw new BlockedError(BlockedReason.unknown, httpCacheItem.policy.responseHeaders());
    }

    if (httpCacheItem.status === 429) {
      const retryAfter = parseInt(`${httpCacheItem.policy.responseHeaders()['retry-after']}`);
      if (!isNaN(retryAfter)) {
        this.rateLimitedCache.set(url.host, undefined, { ttl: retryAfter * 1000 });
      }

      throw new TooManyRequestsError(retryAfter);
    }

    throw new HttpError(httpCacheItem.status, httpCacheItem.statusText, responseHeaders);
  };

  private determineCacheKey(url: URL, init?: CustomRequestInit): string {
    return `${url.href}_${init?.body?.toString()}`;
  }

  private determineCacheTtl(httpCacheItem: HttpCacheItem): number {
    if (httpCacheItem.status === 200) {
      return Math.max(httpCacheItem.policy.timeToLive(), this.MIN_CACHE_TTL);
    }

    return httpCacheItem.policy.timeToLive();
  };

  private cacheGet(key: string): HttpCacheItem | undefined {
    const buffer = this.httpCache.get(key);

    return buffer ? JSON.parse(gunzipSync(buffer).toString()) : undefined;
  }

  private cacheSet(key: string, httpCacheItem: HttpCacheItem) {
    const ttl = this.determineCacheTtl(httpCacheItem);
    if (ttl <= 0) {
      return;
    }

    this.httpCache.set(key, gzipSync(JSON.stringify(httpCacheItem)), { ttl });
  }

  private headersToObject(headers: Headers): Record<string, string> {
    const obj: Record<string, string> = {};

    headers.forEach((value, name) => {
      obj[name] = value;
    });

    return obj;
  };

  private async cachedFetch(ctx: Context, url: URL, init?: CustomRequestInit): Promise<HttpCacheItem> {
    const newInit = this.getInit(ctx, url, init);

    const request: CachePolicy.Request = { url: url.href, method: newInit.method ?? 'GET', headers: {} };

    const cacheKey = this.determineCacheKey(url, init);
    let httpCacheItem = this.cacheGet(cacheKey);
    if (httpCacheItem) {
      this.logger.info(`Cached fetch ${request.method} ${url}`, ctx);
      return this.handleHttpCacheItem(ctx, httpCacheItem, url, init);
    }

    const response = await this.queuedFetch(ctx, url, newInit);
    const body = await response.text();

    const policy = new CachePolicy(request, { status: response.status, headers: this.headersToObject(response.headers) }, { shared: false });

    httpCacheItem = { policy, status: response.status, statusText: response.statusText, body };

    this.cacheSet(cacheKey, httpCacheItem);

    return this.handleHttpCacheItem(ctx, httpCacheItem, url, init);
  };

  protected async fetchWithTimeout(ctx: Context, url: URL, init?: CustomRequestInit): Promise<Response> {
    this.logger.info(`Fetch ${init?.method ?? 'GET'} ${url}`, ctx);

    if (this.rateLimitedCache.has(url.host)) {
      throw new TooManyRequestsError(this.rateLimitedCache.getRemainingTTL(url.host) / 1000);
    }

    if ((this.timeoutsCountCache.get(url.host) ?? 0) >= (init?.timeoutsCountThrow ?? this.DEFAULT_FAILED_REQUEST_COUNT_THROW)) {
      throw new TooManyTimeoutsError();
    }

    let response;
    try {
      response = await fetch(url, { ...init, keepalive: true, signal: AbortSignal.timeout(init?.timeout ?? this.DEFAULT_TIMEOUT) });
    } catch (error) {
      if (error instanceof DOMException && error.name === 'TimeoutError') {
        await this.increaseTimeoutsCount(url);
        throw new TimeoutError();
      }

      throw error;
    }

    await this.decreaseTimeoutsCount(url);

    return response;
  };

  private async increaseTimeoutsCount(url: URL) {
    await this.timeoutsCountMutex.runExclusive(async () => {
      const count = (this.timeoutsCountCache.get(url.host) ?? 0) + 1;
      this.timeoutsCountCache.set(url.host, count, { ttl: this.TIMEOUT_CACHE_TTL });
    });
  }

  private async decreaseTimeoutsCount(url: URL) {
    await this.timeoutsCountMutex.runExclusive(async () => {
      const count = Math.max(0, (this.timeoutsCountCache.get(url.host) ?? 0) - 1);
      this.timeoutsCountCache.set(url.host, count, { ttl: this.TIMEOUT_CACHE_TTL });
    });
  }

  private getSemaphore(url: URL, queueLimit: number, queueTimeout: number): SemaphoreInterface {
    let sem = this.semaphores.get(url.host);

    if (!sem) {
      sem = withTimeout(new Semaphore(queueLimit), queueTimeout, new QueueIsFullError());
      this.semaphores.set(url.host, sem);
    }

    return sem;
  }

  private async queuedFetch(ctx: Context, url: URL, init?: CustomRequestInit): Promise<Response> {
    const queueLimit = init?.queueLimit ?? this.DEFAULT_QUEUE_LIMIT;
    const queueTimeout = init?.queueTimeout ?? this.DEFAULT_QUEUE_TIMEOUT;

    const semaphore = this.getSemaphore(url, queueLimit, queueTimeout);

    const [,release] = await semaphore.acquire();

    try {
      return await this.fetchWithTimeout(ctx, url, init);
    } finally {
      release();
    }
  }
}
