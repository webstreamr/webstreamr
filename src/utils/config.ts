import { Extractor } from '../extractor';
import { Config } from '../types';

export const getDefaultConfig = (): Config => {
  return { multi: 'on', en: 'on' };
};

export const showErrors = (config: Config): boolean => 'showErrors' in config;

export const showExternalUrls = (config: Config): boolean => 'includeExternalUrls' in config;

export const hasMultiEnabled = (config: Config): boolean => 'multi' in config;

export const disableExtractorConfigKey = (extractor: Extractor): string => `disableExtractor_${extractor.id}`;

export const isExtractorDisabled = (config: Config, extractor: Extractor): boolean => disableExtractorConfigKey(extractor) in config;
