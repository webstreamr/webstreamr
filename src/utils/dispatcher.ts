import { socksDispatcher } from 'fetch-socks';
import { minimatch } from 'minimatch';
import { Dispatcher, ProxyAgent } from 'undici';
import { Context } from '../types';

export const getProxyForUrl = (ctx: Context, url: URL): URL | undefined => {
  const proxyConfig = ctx.config['proxyConfig'] || process.env['PROXY_CONFIG'];

  if (proxyConfig) {
    for (const rule of proxyConfig.split(',')) {
      const [hostPattern, proxy] = rule.split(/:(.+)/);
      if (!hostPattern || !proxy) {
        throw new Error(`Proxy rule "${rule}" is invalid.`);
      }

      if (hostPattern === '*' || minimatch(url.host, hostPattern)) {
        return proxy === 'false' ? undefined : new URL(proxy);
      }
    }
  } else if (process.env['ALL_PROXY']) {
    return new URL(process.env['ALL_PROXY']);
  }

  return undefined;
};

export const createProxyAgent = (proxyUrl: URL): Dispatcher => {
  if (proxyUrl.protocol === 'socks5:') {
    return socksDispatcher({ type: 5, host: proxyUrl.hostname, port: parseInt(proxyUrl.port) });
  }

  return new ProxyAgent({ uri: proxyUrl.href });
};
