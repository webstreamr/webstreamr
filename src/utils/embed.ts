import { unpack } from 'unpacker';

export class CouldNotFindStreamLinkInEmbedError extends Error {}

export const extractUrlFromPacked = (html: string, linkRegExps: RegExp[]): URL => {
  const evalMatch = html.match(/eval\(function\(p,a,c,k,e,d\).*\)\)/);
  if (!evalMatch) {
    throw new Error(`No p.a.c.k.e.d string found`);
  }

  const unpacked = unpack(evalMatch[0]);

  for (const linkRegexp of linkRegExps) {
    const linkMatch = unpacked.match(linkRegexp);
    if (linkMatch && linkMatch[1]) {
      return new URL('https://' + linkMatch[1].replace(/^(https:)?\/\//, ''));
    }
  }

  throw new CouldNotFindStreamLinkInEmbedError(`Could not find a stream link in embed`);
};
