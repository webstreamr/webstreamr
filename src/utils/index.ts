export * from './embed';
export * from './fetch';
export * from './imdb';
export * from './language';
export * from './log';
export * from './manifest';
export * from './promise';
