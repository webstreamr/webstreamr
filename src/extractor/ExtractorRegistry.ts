import TTLCache from '@isaacs/ttlcache';
import winston from 'winston';
import { Extractor } from './types';
import { Context, Meta, UrlResult } from '../types';
import { Fetcher } from '../utils';
import { DoodStream } from './DoodStream';
import { Dropload } from './Dropload';
import { SuperVideo } from './SuperVideo';
import { ExternalUrl } from './ExternalUrl';
import { NotFoundError } from '../error';

export class ExtractorRegistry {
  private readonly logger: winston.Logger;
  private readonly extractors: Extractor[];
  private readonly urlResultCache: TTLCache<string, UrlResult | undefined>;

  constructor(logger: winston.Logger, fetcher: Fetcher) {
    this.logger = logger;
    this.extractors = [
      new DoodStream(fetcher),
      new Dropload(fetcher),
      new SuperVideo(fetcher),
      new ExternalUrl(fetcher), // fallback extractor which must come last
    ];
    this.urlResultCache = new TTLCache({ max: 1024 });
  }

  readonly handle = async (ctx: Context, url: URL, meta: Meta): Promise<UrlResult | undefined> => {
    let urlResult = this.urlResultCache.get(url.href);
    if (this.urlResultCache.has(url.href)) {
      return urlResult ? { ...urlResult, ttl: this.urlResultCache.getRemainingTTL(url.href) } : undefined;
    }

    const extractor = this.extractors.find(extractor => extractor.supports(ctx, url));
    if (!extractor) {
      return undefined;
    }

    this.logger.info(`Extract stream URL using ${extractor.id} extractor from ${url}`, ctx);

    try {
      urlResult = await extractor.extract(ctx, url, meta);
    } catch (error) {
      if (error instanceof NotFoundError) {
        this.urlResultCache.set(url.href, urlResult, { ttl: extractor.ttl });

        return undefined;
      }

      return {
        url,
        isExternal: true,
        error,
        label: url.host,
        sourceId: `${extractor.id}`,
        ttl: extractor.ttl,
        meta,
      };
    }

    this.urlResultCache.set(url.href, urlResult, { ttl: extractor.ttl });

    return urlResult;
  };
}
