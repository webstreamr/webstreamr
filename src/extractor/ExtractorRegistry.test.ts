import winston from 'winston';
import { ExtractorRegistry } from './ExtractorRegistry';
import { Context, UrlResult } from '../types';
import { Fetcher } from '../utils';
jest.mock('../utils/Fetcher');

const logger = winston.createLogger({ transports: [new winston.transports.Console({ level: 'nope' })] });
// @ts-expect-error No constructor args needed
const fetcher = new Fetcher();
const extractorRegistry = new ExtractorRegistry(logger, fetcher);

describe('ExtractorRegistry', () => {
  const ctx: Context = { id: 'id', ip: '127.0.0.1', config: { de: 'on' } };

  test('returns error result from extractor', async () => {
    const urlResult = await extractorRegistry.handle(ctx, new URL('https://some-url.test'), { countryCode: 'en' });

    expect(urlResult).toMatchSnapshot();
  });

  test('return external URLs by default', async () => {
    const urlResult = await extractorRegistry.handle(ctx, new URL('https://mixdrop.ag/e/3nzwveprim63or6'), { countryCode: 'de' });

    expect(urlResult).toMatchSnapshot();
  });

  test('does not return external URLs if disabled by config', async () => {
    const urlResult = await extractorRegistry.handle({ ...ctx, config: { ...ctx.config, excludeExternalUrls: 'on' } }, new URL('https://mixdrop.ag/e/l7v73zqrfdj19z'), { countryCode: 'de' });

    expect(urlResult).toBeUndefined();
  });

  test('returns from memory cache if possible', async () => {
    const { ttl: ttl1, ...urlResultRest1 } = await extractorRegistry.handle(ctx, new URL('https://dropload.io/lyo2h1snpe5c.html'), { countryCode: 'de' }) as UrlResult;
    const { ttl: ttl2, ...urlResultRest2 } = await extractorRegistry.handle(ctx, new URL('https://dropload.io/lyo2h1snpe5c.html'), { countryCode: 'de' }) as UrlResult;

    expect(urlResultRest1).not.toBeUndefined();
    expect(urlResultRest2).toStrictEqual(urlResultRest1);

    expect(ttl1).not.toBe(undefined);
    expect(ttl2).not.toBe(undefined);
  });

  test('ignores not found errors but caches them', async () => {
    const urlResult1 = await extractorRegistry.handle(ctx, new URL('https://dropload.io/asdfghijklmn.html'), { countryCode: 'de' });
    const urlResult2 = await extractorRegistry.handle(ctx, new URL('https://dropload.io/asdfghijklmn.html'), { countryCode: 'de' });

    expect(urlResult1).toBeUndefined();
    expect(urlResult2).toBeUndefined();
  });
});
