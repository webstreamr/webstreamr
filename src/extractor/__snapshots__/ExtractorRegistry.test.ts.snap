// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExtractorRegistry return external URLs by default 1`] = `
{
  "isExternal": true,
  "label": "mixdrop.ag",
  "meta": {
    "countryCode": "de",
  },
  "sourceId": "external_de",
  "ttl": 3600000,
  "url": "https://mixdrop.ag/e/3nzwveprim63or6",
}
`;

exports[`ExtractorRegistry returns error result from extractor 1`] = `
{
  "error": [Error: TypeError: fetch failed],
  "isExternal": true,
  "label": "some-url.test",
  "meta": {
    "countryCode": "en",
  },
  "sourceId": "external",
  "ttl": 3600000,
  "url": "https://some-url.test/",
}
`;
